File: arrays-and-hashing/05/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : nums) {
            int frequency = frequencyMap.getOrDefault(num, 0);
            frequencyMap.put(num, frequency + 1);
        }

        List<Entry<Integer, Integer>> frequencyPairs = new ArrayList<>(frequencyMap.entrySet());
        frequencyPairs.sort((e1, e2) -> e2.getValue() - e1.getValue());

        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = frequencyPairs.get(i).getKey();
        }

        return result;
    }
}

File: arrays-and-hashing/05/solution.py
import heapq
from typing import *


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequencies_map = {}
        for num in nums:
            frequencies_map[num] = frequencies_map.get(num, 0) + 1
        kheap = []
        for num, freq in frequencies_map.items():
            if len(kheap) < k:
                heapq.heappush(kheap, (freq, num))
            else:
                if freq > kheap[0][0]:
                    heapq.heappop(kheap)
                    heapq.heappush(kheap, (freq, num))
        return [x for _, x in kheap]
